### PR TITLE
CXX-691 Find options take a hint

### PR DESCRIPTION
**Pull request by samantharitter**

### CODE REVIEW

File: src/mongocxx/collection.cpp [hunk 1]
```diff
@@ -125,6 +124,10 @@ cursor collection::find(bsoncxx::document::view filter, const options::find& opt
         filter_builder << "$orderby" << bsoncxx::types::b_document{*options.sort()};
     }
 
+    if (options.hint()) {
+        filter_builder << builder::stream::concatenate{options.hint()->to_document()};
```
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: This pattern that we use throughout the codebase is bad, because to_document() returns a temporary value that concatenate{} converts to a view and drops on the floor.  View_or_value will fix it!

File: src/mongocxx/options/find.hpp [hunk 2]
```diff
@@ -113,6 +113,21 @@ class MONGOCXX_API find {
     const stdx::optional<enum cursor_type>& cursor_type() const;
 
     ///
+    /// Sets the index to use for this operation.
+    ///
+    /// @param index_hint
+    ///   Object representing the index to use.
+    ///
```
Comment (Machyne) [inline src/mongocxx/options/find.hpp]: Add `@see https://docs.mongodb.org/manual/reference/operator/meta/hint/`.
Should we add "When an index filter exists for the query shape, MongoDB ignores the $hint." as a note?

File: src/mongocxx/test/collection_mocked.cpp [hunk 3]
```diff
@@ -297,22 +297,31 @@ TEST_CASE("Collection", "[collection]") {
...
             REQUIRE(coll == mongo_coll.implementation());
             REQUIRE(flags == MONGOC_QUERY_NONE);
             REQUIRE(skip == skip);
             REQUIRE(limit == limit);
             REQUIRE(batch_size == batch_size);
+
             bsoncxx::document::view query_view{bson_get_data(query), query->len};
-            REQUIRE(query_view["$query"].get_document() == doc);
+
+            REQUIRE(bsoncxx::to_json(query_view["$query"].get_document()) == bsoncxx::to_json(doc));
```
Comment (Machyne) [inline src/mongocxx/test/collection_mocked.cpp]: Why was this changed? Did it not work before?

File: src/mongocxx/options/find.hpp (hunk 2 above)
Comment (acmorrow) [inline src/mongocxx/options/find.hpp]: Yeah, I like the comment suggestion.

File: src/mongocxx/options/find.hpp [hunk 4]
```diff
@@ -30,7 +31,6 @@ namespace options {
 /// Class representing the optional arguments to a MongoDB query.
 ///
 class MONGOCXX_API find {
-
```
Comment (acmorrow) [inline src/mongocxx/options/find.hpp]: Add this back please - we try to add a blank line between the class open and the leading public.

**Comment by acmorrow** [comment]
LGTM

**Comment by samantharitter** [comment]
@acmorrow @tessavitabile this is ready for re-review, it has been updated to use view_or_value.

**Comment by tessavitabile** [comment]
LGTM

File: src/bsoncxx/view_or_value.hpp [hunk 5]
```diff
@@ -123,6 +123,30 @@ class view_or_value {
     View _view;
 };
 
+///
+/// Equality operators for comparison with plain Views.
+///
+
+template <typename View, typename Value>
+BSONCXX_INLINE bool operator==(view_or_value<View, Value> lhs, View rhs) {
```
Comment (acmorrow) [inline src/bsoncxx/view_or_value.hpp]: Given that view_or_value may transfer ownership, I'm very certain you want to say `const view_or_value<View, Value>&` here. For the `View` type argument, I don't think it matters.

**Comment by samantharitter** [comment]
@acmorrow I saw your LGTM from November 28 and mistakenly merged but I'll add any changes to a new commit!

**Comment by acmorrow** [comment]
I think that is the only one. I maybe missed it last time.

**Comment by samantharitter** [comment]
Nope, you didn't miss it, that change is new

File: src/bsoncxx/view_or_value.hpp (hunk 5 above)
Comment (samantharitter) [inline src/bsoncxx/view_or_value.hpp]: That makes sense.  The same should be true for the operator== below where view_or_value is on the rhs, I think.

Comment (acmorrow) [inline src/bsoncxx/view_or_value.hpp]: Right. All of them need it.
